Test shape analysis with opinfos

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59814 Test shape analysis with opinfos**
* #59812 add ssupport for a few more opinfos in jit
* #59811 Don't substitute in symbolic shapes to shape compute graph
* #59810 small cleanups
* #59809 Only optimize after change (redo)

Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

Commit:
Test shape analysis with opinfos

[ghstack-poisoned]

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -456,6 +457,19 @@ def get_sample():
                         check_alias_annotation(name, (get_sample(),) + sample.args, sample.kwargs,
                                                func_type=func_type, aten_name=op.aten_name)
 
+                        # TODO: use script graph as well
```
The dtype check above is incorrect. First, this test only runs in float32 and complexfloat32. See the definition of `_variant_ops` here: 

https://github.com/pytorch/pytorch/blob/0ceea7faf4f864ff834617150f1d58d9c1c56b2b/test/test_ops.py#L267

Second, the comment says it would only check if the dtype is float32 or int64 but the check is actually for float32 and int32.

eellison replied:
Okay, updated to say, only running in float32..

mruberry commented on test/test_ops.py:
```diff
@@ -456,6 +457,19 @@ def get_sample():
+                                                      *sample.args,
+                                                      **sample.kwargs)
+
+                            if isinstance(out, torch.Tensor):
```
Probably want a TODO here to handle ops that output multiple tensors?

eellison replied:
will do, there also isn't support for shape analysis functions with multiple outputs either..

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -236,8 +236,9 @@ def __init__(self,
-                 has_fake_function=False  # does this opinfo use a function to simulate functional behavior
+                 has_fake_function=False,  # does this opinfo use a function to simulate functional behavior
                                             # of a mutating op
+                 assert_jit_shape_analysis = False,  # assert that jit shape analysis fully propagates shape
```
Will people writing OpInfos know to set this properly?

eellison replied:
i'm not really expecting people to add shape analysis at least initially, it's still very early in implementation and has a few rough edges

mruberry replied:
OK, as the long as the default is false and it won't yell at people

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

mruberry commented:
OpInfo-related changes look good in this PR but see comments in #59812 about a couple tweaks that should make the changes even simpler

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -1098,7 +1098,7 @@ bool TensorType::isSubtypeOfExt(const TypePtr& rhs, std::ostream* why_not) const
     if (this == rhs_p.get()) {
       return true;
     }
-    return *merge(*rhs_p) == *rhs_p;
```
cc @Krovatkin for review. the logic here was wrong previously. If you have:
RHS: Tensor(15, 5)
This: Tensor
Merge: Tensor

Then this subtypes RHS because the merge is equal to current type. we were checking if rhs subtypes this previously

eellison commented on torch/csrc/jit/passes/shape_analysis.cpp:
```diff
@@ -2176,7 +2176,7 @@ using TypeCache = std::unordered_map<TypePtr, TypePtr>;
 TypePtr getOrCreateUnshapedType(TypePtr type, TypeCache& unshaped_type_cache);
 
 TypePtr unshapedTypeImpl(TypePtr type, TypeCache& unshaped_type_cache) {
-  if (type->isSubtypeOf(TensorType::get())) {
```
this worked bc of our incorrect other logic lol...

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

eellison commented:
@mruberry there may be changes on the other PR but I think this one is ready to go

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃 

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).

Commit:
Update on "Test shape analysis with opinfos"


Using opinfos to test shape analysis. By default, we just check that we don't give incorrect answers, and then if `assert_jit_shape_analysis` is true, tests that we correctly propagates the full shape. and it found a couple bugs 😃

Differential Revision: [D30200058](https://our.internmc.facebook.com/intern/diff/D30200058)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30200058).
